### PR TITLE
feat: add `resourceType` support to `cy.intercept()` on `req`/`routeMatcher`

### LINKED ISSUE #14525: cy.intercept by request cause
### What would you like?

Allow matching routes by its cause (e. g. `image` or `xmlhttprequest`):

```js
cy.intercept({
  url: "/orders",
  cause: "xmlhttprequest",
});
```

### Why is this needed?

My current application uses same URLs on API and UI side (e. g. `https://127.0.0.1:3000/orders` on the frontend queries `https://127.0.0.1:4000/orders` on the backend). I need to stub the API call, then open the page on the frontend as usual.

`cy.route` was handling only API requests, so simple `cy.route("/orders")` would work alright independent of the environment – if the backend was on a different hostname (e. g. CI could use a staging instance instead of localhost), `cy.route` would still match it. With `cy.intercept`, the only way to match API calls right now is to add a hostname, e. g.:

```js
const apiRoot = Cypress.env("api_root");
cy.intercept("${apiRoot}/orders");
```

This looks acceptable but it's another configuration variable to keep up to date, and can be avoided. :-)

### Implementation suggestion

If I get this correctly, this can't be implemented using just the current setup with running a proxy server. However, we could use the extension running in the browser to match the routes we need using the `webRequest.onBeforeRequest` ([Mozdev][mdn], [Chrome docs][chr]) API. If we can't move all intercepting business to the extension (why?), we could use it to add an HTTP header to all requests (`X-Cypress-Cause: xmlhttprequest`) and match that on the proxy side as usual.

P. S. Thank you all for your great work!

[mdn]: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeRequest
[chr]: https://developer.chrome.com/docs/extensions/reference/webRequest/

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #14525

### User facing changelog

- Added the ability to match on `resourceType` with `cy.intercept()`, and to see the resource type of an intercepted request as `req.resourceType`.

### Additional details

- All browsers report this in pre-request information to enable proxy logging, so we can now safely add it in the public `intercept` API too.
- In combination with `cy.intercept(..., { log: false }`, these two changes will unblock users on disabling/enabling logs by resource type: https://github.com/cypress-io/cypress/issues/9358 (no PR yet for 9358)
- Also renamed "resourceType" to "requestedWith" everywhere it's used in multi-domain. Multi-domain was previously using the term "resource type" to refer to "xhr" or "fetch" for the runner patches, but this usage of the term clobbers the meaning given by CDP and makes it confusing. Since it's referred to as "requestedWith" in some places in MD already, I went ahead and renamed MD's usage to "requestedWith" for the entirety of MD.

### Steps to test

Use the new `resourceType` property on `RouteMatcher` and `req` to match only certain requests:

```ts
cy.intercept({ resourceType: 'stylesheet' }) // only CSS

cy.intercept('/my-api', (req) => {
  expect(req.resourceType).to.eq('xhr') // only XHRs should hit the API
})
```

Possible values:

```ts
type ResourceType = 'document' | 'fetch' | 'xhr' | 'websocket' | 'stylesheet' | 'script' | 'image' | 'font' | 'cspviolationreport' | 'ping' | 'manifest' | 'other'
```

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4921
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/app/cypress.config.ts [hunk 1]
```diff
@@ -38,7 +38,6 @@ export default defineConfig({
       delete process.env.CYPRESS_INTERNAL_E2E_TESTING_SELF_PARENT_PROJECT
       process.env.CYPRESS_INTERNAL_E2E_TESTING_SELF = 'true'
       process.env.CYPRESS_INTERNAL_VITE_OPEN_MODE_TESTING = 'true'
-      // process.env.DEBUG = '*'
```
Comment (flotwig) [inline packages/app/cypress.config.ts:41]: unrelated change, just noticed this while wrestling with `debug`

**Review by chrisbreiding — APPROVED** [review]

File: packages/driver/cypress/e2e/commands/net_stubbing.cy.ts [hunk 2]
```diff
@@ -3568,17 +3546,72 @@ describe.skip('network stubbing', function () {
...
+        .then((win) => {
+          const controller = new AbortController()
+          const { signal } = controller
+
+          fetch('/users/', { signal, method: 'POST' }).catch((e) => {
+          // do nothing on an abort
+          })
+
+          // if you abort too fast in firefox or safari, the fetch is never sent to the server for us to intercept
+          if (!Cypress.isBrowser('chrome')) {
```
Comment (AtofStryker) [inline packages/driver/cypress/e2e/commands/net_stubbing.cy.ts]: is this an issue in electron or edge?
```suggestion
          if (!Cypress.isBrowser({ family: 'chromium' })) {
```

**Review by AtofStryker — APPROVED** [review]
Looks good on my end! Thank you for renaming `resourceType` to `requestedWith`. I gave this a run against some of the tests in cypress-origin-providers and all checks out

File: packages/driver/cypress/e2e/commands/net_stubbing.cy.ts (hunk 2 above)
Comment (flotwig) [reply packages/driver/cypress/e2e/commands/net_stubbing.cy.ts]: Hmm, yeah, `{ family: 'chromium' }` seems like it could be right. @mjhenkes thoughts?

Comment (mjhenkes) [reply packages/driver/cypress/e2e/commands/net_stubbing.cy.ts]: chromium should be fine here. I don't think this is an issue in edge or electron
